lib/Dialect/SV/Transforms/SVExtractTestCode.cpp

```diff
@@ -0,0 +1,224 @@
+static void addBlockMapping(BlockAndValueMapping &cutMap, Operation *oldOp,
+                            Operation *newOp) {
+  assert(oldOp->getNumRegions() == newOp->getNumRegions());
+  for (size_t i = 0, e = oldOp->getNumRegions(); i != e; ++i) {
+    auto &oldRegion = oldOp->getRegion(i);
+    auto &newRegion = newOp->getRegion(i);
+    for (auto oi = oldRegion.begin(), oe = oldRegion.end(); oi != oe; ++oi) {
+      cutMap.map(&*oi, &newRegion.emplaceBlock());
+    }
+  }
+}
+
+// Do the cloning, which is just a pre-order traversal over the module looking
+// for marked ops.
+static void migrateOps(hw::HWModuleOp oldMod, hw::HWModuleOp newMod,
+                       SetVector<Operation *> &depOps,
+                       BlockAndValueMapping &cutMap) {
+  OpBuilder b = OpBuilder::atBlockBegin(newMod.getBodyBlock());
+  oldMod.walk<WalkOrder::PreOrder>([&](Operation *op) {
+    if (depOps.count(op)) {
+      setInsertPointToEndOrTerminator(b, cutMap.lookup(op->getBlock()));
+      auto newOp = b.cloneWithoutRegions(*op, cutMap);
+      addBlockMapping(cutMap, op, newOp);
+    }
+  });
+}
+
+//===----------------------------------------------------------------------===//
+// StubExternalModules Pass
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+struct SVExtractTestCodeImplPass
+    : public SVExtractTestCodeBase<SVExtractTestCodeImplPass> {
+  void runOnOperation() override;
+
+private:
+  void doModule(hw::HWModuleOp module, std::function<bool(Operation *)> fn,
+                StringRef suffix, StringRef path) {
+    // Find Operations of interest.
+    SmallVector<Operation *> roots;
+    module->walk([&fn, &roots](Operation *op) {
+      if (fn(op))
+        roots.push_back(op);
+    });
+    // No Ops?  No problem.
+    if (roots.empty())
+      return;
+    // Find the data-flow and structural ops to clone.  Result includes roots.
+    auto opsToClone = computeCloneSet(roots);
+    // Find the dataflow into the clone set
+    SetVector<Value> inputs;
+    for (auto op : opsToClone)
+      for (auto arg : op->getOperands()) {
+        auto argOp = arg.getDefiningOp(); // may be null
+        if (!opsToClone.count(argOp))
+          inputs.insert(arg);
+      }
+
+    // Make a module to contain the clone set, with arguments being the cut
+    BlockAndValueMapping cutMap;
+    auto bmod = createModuleForCut(module, inputs, cutMap, suffix, path);
+    // do the clone
+    migrateOps(module, bmod, opsToClone, cutMap);
+    // erase old operations of interest
+    for (auto op : roots)
+      op->erase();
+  }
+};
+
+} // end anonymous namespace
+
+void SVExtractTestCodeImplPass::runOnOperation() {
+  auto *topLevelModule = getOperation().getBody();
+  for (auto &op : topLevelModule->getOperations())
+    if (auto rtlmod = dyn_cast<hw::HWModuleOp>(op)) {
+      // Extract two sets of ops to different modules
+      doModule(
+          rtlmod,
+          [](Operation *op) -> bool {
```

Comment:
nit, but I'd suggest pulling this out to a named lambda to improve formatting:

auto isAssert = [](...);